// @mantine/core/esm/components/Menubar/MenubarTarget/MenubarTarget.mjs
"use client";
import { createEventHandler } from "../../../core/utils/create-event-handler/create-event-handler.mjs";
import { useProps } from "../../../core/MantineProvider/use-props/use-props.mjs";
import { useDirection } from "../../../core/DirectionProvider/DirectionProvider.mjs";
import { UnstyledButton } from "../../UnstyledButton/UnstyledButton.mjs";
import { Popover } from "../../Popover/Popover.mjs";
import { useMenubarContext, useMenubarMenuContext } from "../Menubar.context.mjs";
import { jsx } from "react/jsx-runtime";
//#region packages/@mantine/core/src/components/Menubar/MenubarTarget/MenubarTarget.tsx
const defaultProps = { refProp: "ref" };
function isPrintableKey(event) {
	return event.key.length === 1 && event.key !== " " && !event.ctrlKey && !event.metaKey && !event.altKey;
}
function getTargetLabel(node) {
	return (node.textContent ?? "").trim().toLowerCase();
}
function MenubarTarget(props) {
	const { children, refProp, className, style, disabled, "data-disabled": dataDisabled, onClick, onMouseEnter, onMouseLeave, onFocus, onKeyDown, ...others } = useProps("MenubarTarget", defaultProps, props);
	const ctx = useMenubarContext();
	const menuCtx = useMenubarMenuContext();
	const { dir } = useDirection();
	const isDisabled = disabled || dataDisabled;
	const isActive = ctx.activeIndex === menuCtx.index;
	const isUnresolvedTabStop = menuCtx.index === -1 && !isDisabled;
	const moveToAdjacent = (direction) => {
		const nextIndex = ctx.getAdjacentIndex(menuCtx.index, direction);
		ctx.setActiveIndex(nextIndex);
		ctx.focusTarget(nextIndex);
		if (ctx.openIndex !== null) ctx.openMenu(nextIndex, "click");
	};
	const openAndFocusItem = (itemPosition) => {
		ctx.openMenu(menuCtx.index, "click");
		ctx.setActiveIndex(menuCtx.index);
		ctx.focusMenuItem(menuCtx.index, itemPosition);
	};
	const typeAhead = (event) => {
		const targets = ctx.getTargets();
		if (targets.length === 0) return;
		const char = event.key.toLowerCase();
		for (let offset = 1; offset <= targets.length; offset += 1) {
			const index = (menuCtx.index + offset) % targets.length;
			const node = targets[index];
			if (node && !node.disabled && !node.hasAttribute("data-disabled") && getTargetLabel(node).startsWith(char)) {
				event.preventDefault();
				ctx.setActiveIndex(index);
				node.focus();
				if (ctx.openIndex !== null) ctx.openMenu(index, "click");
				break;
			}
		}
	};
	const handleClick = createEventHandler(onClick, () => {
		if (isDisabled) return;
		if (ctx.openIndex === menuCtx.index && ctx.getOpenSource() !== "hover") ctx.closeMenu();
		else {
			ctx.openMenu(menuCtx.index, "click");
			ctx.setActiveIndex(menuCtx.index);
		}
	});
	const handleMouseEnter = createEventHandler(onMouseEnter, () => {
		if (isDisabled) return;
		ctx.cancelClose();
		if (ctx.trigger === "hover") {
			ctx.openMenu(menuCtx.index, "hover");
			ctx.setActiveIndex(menuCtx.index);
		} else if (ctx.openIndex !== null && ctx.openIndex !== menuCtx.index) {
			ctx.openMenu(menuCtx.index, "hover");
			ctx.setActiveIndex(menuCtx.index);
		}
	});
	const handleMouseLeave = createEventHandler(onMouseLeave, () => {
		if (ctx.trigger === "hover") ctx.scheduleClose();
	});
	const handleFocus = createEventHandler(onFocus, () => ctx.setActiveIndex(menuCtx.index));
	const handleKeyDown = createEventHandler(onKeyDown, (event) => {
		if (isDisabled) return;
		const forwardKey = dir === "rtl" ? "ArrowLeft" : "ArrowRight";
		const backKey = dir === "rtl" ? "ArrowRight" : "ArrowLeft";
		if (event.key === forwardKey) {
			event.preventDefault();
			moveToAdjacent(1);
		} else if (event.key === backKey) {
			event.preventDefault();
			moveToAdjacent(-1);
		} else if (event.key === "ArrowDown" || event.key === "Enter" || event.key === " ") {
			event.preventDefault();
			openAndFocusItem("first");
		} else if (event.key === "ArrowUp") {
			event.preventDefault();
			openAndFocusItem("last");
		} else if (event.key === "Home") {
			event.preventDefault();
			const enabled = ctx.getEnabledIndexes();
			if (enabled.length > 0) {
				ctx.setActiveIndex(enabled[0]);
				ctx.focusTarget(enabled[0]);
			}
		} else if (event.key === "End") {
			event.preventDefault();
			const enabled = ctx.getEnabledIndexes();
			if (enabled.length > 0) {
				const last = enabled[enabled.length - 1];
				ctx.setActiveIndex(last);
				ctx.focusTarget(last);
			}
		} else if (event.key === "Escape") {
			if (menuCtx.opened) ctx.closeMenu();
		} else if (event.key === "Tab") {
			if (menuCtx.opened) ctx.closeMenu();
		} else if (isPrintableKey(event)) typeAhead(event);
	});
	return /* @__PURE__ */ jsx(Popover.Target, {
		refProp,
		popupType: "menu",
		children: /* @__PURE__ */ jsx(UnstyledButton, {
			...others,
			unstyled: ctx.unstyled,
			role: "menuitem",
			tabIndex: isActive || isUnresolvedTabStop ? 0 : -1,
			disabled,
			"data-menubar-target": true,
			"data-menubar-id": menuCtx.id,
			"data-disabled": dataDisabled || void 0,
			"data-expanded": menuCtx.opened || void 0,
			"data-mantine-stop-propagation": menuCtx.opened || void 0,
			...ctx.getStyles("target", {
				className,
				style
			}),
			onClick: handleClick,
			onMouseEnter: handleMouseEnter,
			onMouseLeave: handleMouseLeave,
			onFocus: handleFocus,
			onKeyDown: handleKeyDown,
			children
		})
	});
}
MenubarTarget.displayName = "@mantine/core/MenubarTarget";
//#endregion
export { MenubarTarget };
