// @mantine/core/esm/components/DataList/DataList.mjs
"use client";
import { getFontSize, getLineHeight, getSpacing } from "../../core/utils/get-size/get-size.mjs";
import { createVarsResolver } from "../../core/styles-api/create-vars-resolver/create-vars-resolver.mjs";
import { useProps } from "../../core/MantineProvider/use-props/use-props.mjs";
import { useStyles } from "../../core/styles-api/use-styles/use-styles.mjs";
import { factory } from "../../core/factory/factory.mjs";
import { Box } from "../../core/Box/Box.mjs";
import { DataListProvider } from "./DataList.context.mjs";
import DataList_module_default from "./DataList.module.mjs";
import { DataListItem } from "./DataListItem/DataListItem.mjs";
import { DataListItemLabel } from "./DataListItemLabel/DataListItemLabel.mjs";
import { DataListItemValue } from "./DataListItemValue/DataListItemValue.mjs";
import { jsx } from "react/jsx-runtime";
//#region packages/@mantine/core/src/components/DataList/DataList.tsx
const defaultProps = { orientation: "horizontal" };
const varsResolver = createVarsResolver((_, { size, gap, labelWidth }) => ({ root: {
	"--data-list-fz": getFontSize(size),
	"--data-list-lh": getLineHeight(size),
	"--data-list-gap": getSpacing(gap),
	"--data-list-label-width": labelWidth !== void 0 ? typeof labelWidth === "number" ? `${labelWidth}px` : labelWidth : void 0
} }));
const DataList = factory((_props) => {
	const props = useProps("DataList", defaultProps, _props);
	const { classNames, className, style, styles, unstyled, vars, children, gap, orientation, withDivider, labelWidth, mod, attributes, ...others } = props;
	const getStyles = useStyles({
		name: "DataList",
		classes: DataList_module_default,
		props,
		className,
		style,
		classNames,
		styles,
		unstyled,
		attributes,
		vars,
		varsResolver
	});
	return /* @__PURE__ */ jsx(DataListProvider, {
		value: { getStyles },
		children: /* @__PURE__ */ jsx(Box, {
			...getStyles("root"),
			component: "dl",
			mod: [{
				orientation,
				"with-divider": withDivider
			}, mod],
			...others,
			children
		})
	});
});
DataList.classes = DataList_module_default;
DataList.varsResolver = varsResolver;
DataList.displayName = "@mantine/core/DataList";
DataList.Item = DataListItem;
DataList.ItemLabel = DataListItemLabel;
DataList.ItemValue = DataListItemValue;
//#endregion
export { DataList };
