// @mantine/core/cjs/components/DataList/DataList.cjs
"use client";
const require_get_size = require("../../core/utils/get-size/get-size.cjs");
const require_create_vars_resolver = require("../../core/styles-api/create-vars-resolver/create-vars-resolver.cjs");
const require_use_props = require("../../core/MantineProvider/use-props/use-props.cjs");
const require_use_styles = require("../../core/styles-api/use-styles/use-styles.cjs");
const require_factory = require("../../core/factory/factory.cjs");
const require_Box = require("../../core/Box/Box.cjs");
const require_DataList_context = require("./DataList.context.cjs");
const require_DataList_module = require("./DataList.module.cjs");
const require_DataListItem = require("./DataListItem/DataListItem.cjs");
const require_DataListItemLabel = require("./DataListItemLabel/DataListItemLabel.cjs");
const require_DataListItemValue = require("./DataListItemValue/DataListItemValue.cjs");
let react_jsx_runtime = require("react/jsx-runtime");
//#region packages/@mantine/core/src/components/DataList/DataList.tsx
const defaultProps = { orientation: "horizontal" };
const varsResolver = require_create_vars_resolver.createVarsResolver((_, { size, gap, labelWidth }) => ({ root: {
	"--data-list-fz": require_get_size.getFontSize(size),
	"--data-list-lh": require_get_size.getLineHeight(size),
	"--data-list-gap": require_get_size.getSpacing(gap),
	"--data-list-label-width": labelWidth !== void 0 ? typeof labelWidth === "number" ? `${labelWidth}px` : labelWidth : void 0
} }));
const DataList = require_factory.factory((_props) => {
	const props = require_use_props.useProps("DataList", defaultProps, _props);
	const { classNames, className, style, styles, unstyled, vars, children, gap, orientation, withDivider, labelWidth, mod, attributes, ...others } = props;
	const getStyles = require_use_styles.useStyles({
		name: "DataList",
		classes: require_DataList_module.default,
		props,
		className,
		style,
		classNames,
		styles,
		unstyled,
		attributes,
		vars,
		varsResolver
	});
	return /* @__PURE__ */ (0, react_jsx_runtime.jsx)(require_DataList_context.DataListProvider, {
		value: { getStyles },
		children: /* @__PURE__ */ (0, react_jsx_runtime.jsx)(require_Box.Box, {
			...getStyles("root"),
			component: "dl",
			mod: [{
				orientation,
				"with-divider": withDivider
			}, mod],
			...others,
			children
		})
	});
});
DataList.classes = require_DataList_module.default;
DataList.varsResolver = varsResolver;
DataList.displayName = "@mantine/core/DataList";
DataList.Item = require_DataListItem.DataListItem;
DataList.ItemLabel = require_DataListItemLabel.DataListItemLabel;
DataList.ItemValue = require_DataListItemValue.DataListItemValue;
//#endregion
exports.DataList = DataList;
